gh-94943: [Enum] improve repr() when inheriting from a dataclass (#99740)
python/cpython: Doc/howto/enum.rst

(See `OrderedEnum`_ for an example.)

> .. _dataclass_support:

Reviewer: ```suggestion
.. _enum-dataclass-support:
```

To follow standard conventions used in this document and generally

Dataclass support
-----------------

> When inheriting from a :func:`dataclass` the :func:`repr` omits the inherited
> class' name.  For example::

Reviewer: ```suggestion
When inheriting from a :class:`~dataclasses.dataclass`,
the :meth:`~Enum.__repr__` omits the inherited class' name.  For example::
```

* Fix the cross-reference to dataclass
* Format it as a class/object rather than as a callable
* Reference the Enum specific `__repr__` rather than the general `repr` builtin
* Add comma for readibility

    >>> @dataclass
    ... class CreatureDataMixin:
    ...     size: str
    ...     legs: int
    ...     tail: bool = field(repr=False, default=True)
    ...
    >>> class Creature(CreatureDataMixin, Enum):
    ...     BEETLE = 'small', 6
    ...     DOG = 'medium', 4
    ...

>     >>> Creature.DOG
>     <Creature.DOG: size='medium', legs=4>

Reviewer: ```suggestion
    >>> repr(Creature.DOG)
    "<Creature.DOG: size='medium', legs=4>"
```

Maybe a little clearer to explicitly call `repr`, rather than just having the interactive interpreter implicitly print the repr()? Up to you, of course.

Author: I"m going to leave this one as-is as I think it matches the rest of the enum documentation better.

> Use the dataclass option ``repr=False`` to use the standard :func:`repr`.

Reviewer: ```suggestion
Use the :func:`!dataclass` argument ``repr=False``
to use the standard :func:`repr`.
```

* Use standard Python terminology
* Use roles to clarity we're referring to a decorator argument here
